File: 3225-length-of-longest-subarray-with-at-most-k-frequency/length-of-longest-subarray-with-at-most-k-frequency.py
class Solution:
    def maxSubarrayLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        freq = defaultdict(int)
        L = 0
        res = 0
        over_k = 0

        for R in range(n):
            freq[nums[R]] += 1

            if freq[nums[R]] == k + 1:
                over_k += 1
            if over_k:
                freq[nums[L]] -= 1
                if freq[nums[L]] == k:
                    over_k -= 1
                L += 1
        return n - L

File: 3225-length-of-longest-subarray-with-at-most-k-frequency/length-of-longest-subarray-with-at-most-k-frequency.cpp
class Solution {
public:
    int maxSubarrayLength(vector<int>& nums, int k) {
        int L = 0, res = 0;
        unordered_map<int, int> mp;
        for (int R = 0; R < nums.size(); ++R) {
            mp[nums[R]]++;

            while (mp[nums[R]] > k) {
                mp[nums[L]]--;
                L++;
            }

            res = max(res, R - L + 1);
        }

        return res;
    }
};